### PR TITLE
[Uuid] Added the polyfill

### PR DESCRIPTION
**Pull request by lyrixx**

### CODE REVIEW

File: src/Uuid/Uuid.php [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            return self::UUID_VARIANT_MICROSOFT;
+        }
+
+        return self::UUID_VARIANT_OTHER;
+    }
+
+    public static function uuid_time(string $uuid)
+    {
+        if (PHP_INT_SIZE !== 8) {
+            throw new \RuntimeException('UUID time generation is not supported on 32bits system.');
```
Comment (javiereguiluz) [inline src/Uuid/Uuid.php]: `[...] 32bits system.` -> `[...] 32-bit systems.`

Comment (nicolas-grekas) [reply src/Uuid/Uuid.php]: should we suggest to install the extension?

File: src/Util/TestListenerTrait.php [hunk 2]
```diff
@@ -27,6 +27,11 @@ public function startTestSuite($mainSuite)
         $SkippedTestError = class_exists('PHPUnit\Framework\SkippedTestError') ? 'PHPUnit\Framework\SkippedTestError' : 'PHPUnit_Framework_SkippedTestError';
 
         foreach ($mainSuite->tests() as $suite) {
+            // We do not want to test the uuid polyfill on PHP < 7
+            // Since this script try to load the SUT, we must skip it
```
Comment (nicolas-grekas) [inline src/Util/TestListenerTrait.php]: s/try/tries

File: src/Util/TestListenerTrait.php [hunk 3]
```diff
@@ -27,6 +27,11 @@ public function startTestSuite($mainSuite)
...
+            if ('Symfony\Polyfill\Tests\Uuid\UuidTest' === $suite->getName() && version_compare('7.0.0', PHP_VERSION, '>=')) {
```
Comment (nicolas-grekas) [inline src/Util/TestListenerTrait.php]: PHP_VERSION_ID?

File: src/Uuid/Uuid.php (hunk 1 above)
Comment (lyrixx) [reply src/Uuid/Uuid.php]: Thanks. I added the suggestion

File: src/Util/TestListenerTrait.php [hunk 4]
```diff
@@ -109,8 +114,10 @@ function {$f['name']}{$f['signature']}
                 $mainSuite->addTest(new TestListener($suite));
             }
         }
-        foreach ($warnings as $w) {
-            $mainSuite->addTest($w);
+        if (isset($warnings)) {
+            foreach ($warnings as $w) {
```
Comment (dunglas) [inline src/Util/TestListenerTrait.php]: `$warning ?? []`? (nitpicking)

Comment (nicolas-grekas) [reply src/Util/TestListenerTrait.php]: Hello PHP 5.3 :)

File: src/Util/TestListenerTrait.php [hunk 5]
```diff
@@ -27,6 +27,11 @@ public function startTestSuite($mainSuite)
         $SkippedTestError = class_exists('PHPUnit\Framework\SkippedTestError') ? 'PHPUnit\Framework\SkippedTestError' : 'PHPUnit_Framework_SkippedTestError';
 
         foreach ($mainSuite->tests() as $suite) {
+            // We do not want to test the uuid polyfill on PHP < 7
+            // Since this script tries to load the SUT, we must skip it
+            if ('Symfony\Polyfill\Tests\Uuid\UuidTest' === $suite->getName() && PHP_VERSION_ID < 70000) {
```
Comment (stof) [inline src/Util/TestListenerTrait.php]: can we implemenent that in a more generic way by looking at the `@requires php` annotations on the test class instead ?

File: src/Uuid/Uuid.php [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        if (($parsed['clock_seq'] & 0x2000) == 0) {
+            return self::UUID_VARIANT_MICROSOFT;
+        }
+
+        return self::UUID_VARIANT_OTHER;
+    }
+
+    public static function uuid_time(string $uuid)
+    {
+        if (PHP_INT_SIZE !== 8) {
```
Comment (stof) [inline src/Uuid/Uuid.php]: should be fully-qualified to be resolved at compile time and let OPCache optimize dead code due to this condition (same reason that for our PHP version checks)

File: src/Uuid/Uuid.php [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+            // four most significant bits holds version number 4
+            random_int(0, 0x0fff) | 0x4000,
+
+            // 16 bits:
+            // * 8 bits for "clk_seq_hi_res",
+            // * 8 bits for "clk_seq_low",
+            // two most significant bits holds zero and one for variant DCE1.1
+            random_int(0, 0x3fff) | 0x8000,
+
+            // 48 bits for "node"
```
Comment (stof) [inline src/Uuid/Uuid.php]: will this work on 32-bits PHP ?

File: src/Uuid/Uuid.php [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+        );
+    }
+
+    private static function uuid_parse_as_array(string $uuid)
+    {
+        if (36 != \strlen($uuid)) {
+            return null;
+        }
+
+        static $regex = '{^(?<time_low>[0-9a-f]{8})-(?<time_mid>[0-9a-f]{4})-(?<time_hi_and_version>[0-9a-f]{4})-(?<clock_seq>[0-9a-f]{4})-(?<node>[0-9a-f]{12})$}i';
```
Comment (stof) [inline src/Uuid/Uuid.php]: do we even need static variable, or should we inline the regex ?

File: src/Util/TestListenerTrait.php (hunk 5 above)
Comment (stof) [reply src/Util/TestListenerTrait.php]: btw, why do we need to make that filtering here ? Is it only because the polyfill implementation uses PHP 7 typehints ? If yes, it might make more sense to drop them (especially given that the polyfilled functions don't have them).

Comment (lyrixx) [reply src/Util/TestListenerTrait.php]: > can we implemenent that in a more generic way by looking at the @requires php annotations on the test class instead ?

We could, but there is already some harcoded check in this file. More over we don't have an annotation reader in this project.But maybe phpunit expose that information. If it's the case, I will update the code. If not, I prefer to keep something simple

> Is it only because the polyfill implementation uses PHP 7 typehints ? 

yes I guess

> If yes, it might make more sense to drop them (especially given that the polyfilled functions don't have them).

If I do that, I will have to manually check the argument type by hand to have the same behavior than the ext.

File: src/Uuid/Uuid.php (hunk 7 above)
Comment (lyrixx) [reply src/Uuid/Uuid.php]: I don't know. How could I test it ?

File: src/Uuid/Uuid.php [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+        );
+
+        if (null === self::uuid_parse_as_array($uuid)) {
+            return false;
+        }
+
+        return $uuid;
+    }
+
+    private static function uuid_generate_random(): string
```
Comment (stof) [inline src/Uuid/Uuid.php]: isn't it possible for this implementation to generate a null UUID with a non-zero probability ? And is it the same for the extension ?

File: src/Uuid/Uuid.php (hunk 8 above)
Comment (lyrixx) [reply src/Uuid/Uuid.php]: I did it for readability, but indeed :+1: I will move it

File: src/Uuid/Uuid.php (hunk 9 above)
Comment (lyrixx) [reply src/Uuid/Uuid.php]: a null UUID is still a string. Did I miss something?

File: src/Util/TestListenerTrait.php (hunk 5 above)
Comment (stof) [reply src/Util/TestListenerTrait.php]: does the ext actually throw TypeError or warnings ?

File: src/Uuid/Uuid.php (hunk 9 above)
Comment (stof) [reply src/Uuid/Uuid.php]: well, the question is whether `uuid_generate_random()` is allowed to return this null UUID in the original implementation, or whether it never returns it (given that a null UUID won't be of type `UUID_TYPE_RANDOM` AFAICT)

File: src/Util/TestListenerTrait.php (hunk 5 above)
Comment (lyrixx) [reply src/Util/TestListenerTrait.php]: It seems it triggers warning: 
```
$ php -a
php > uuid_create('a');
PHP Warning:  uuid_create() expects parameter 1 to be int, string given in php shell code on line 1

Warning: uuid_create() expects parameter 1 to be int, string given in php shell code on line 1
```

Comment (stof) [reply src/Util/TestListenerTrait.php]: then the polyfill must trigger warnings too

File: src/Uuid/Uuid.php (hunk 9 above)
Comment (lyrixx) [reply src/Uuid/Uuid.php]: It looks like nothing [prevent](https://github.com/karelzak/util-linux/blob/6a735c2c5bfd64b06fd210663c1ebfeac6eef061/libuuid/src/gen_uuid.c#L502-L523) the extension to return a null UUID

**BUT** This is not possible since we "pad" some value for ensure the type:
```
            // 16 bits for "time_hi_and_version",
            // four most significant bits holds version number 4
            random_int(0, 0x0fff) | 0x4000,
```

File: src/Uuid/Uuid.php (hunk 8 above)
Comment (lyrixx) [reply src/Uuid/Uuid.php]: Fixed

File: src/Uuid/Uuid.php (hunk 6 above)
Comment (lyrixx) [reply src/Uuid/Uuid.php]: Fixed. Thanks

File: src/Uuid/Uuid.php (hunk 7 above)
Comment (lyrixx) [reply src/Uuid/Uuid.php]:
```
php > echo random_int(0, 0xffffffffffff);
PHP Warning:  Uncaught TypeError: random_int() expects parameter 2 to be integer, float given in php shell code:1
Stack trace:
#0 php shell code(1): random_int(0, 2.8147497671066E+14)
#1 {main}
  thrown in php shell code on line 1
```

So I updated this :)

File: src/Util/TestListenerTrait.php (hunk 5 above)
Comment (lyrixx) [reply src/Util/TestListenerTrait.php]: OK, I used PHPUnit groups to fixed this issue. I was not able to retrieve requirements :(

**Comment by lyrixx** [comment]
I think this PR is ready :tada:

File: src/Util/TestListenerTrait.php (hunk 5 above)
Comment (stof) [reply src/Util/TestListenerTrait.php]: well, if you change the polyfill to properly trigger warnings rather than TypeErrors, you might not need this at all anymore (as it will become compatible with PHP 5.x too)

**Comment by stof** [comment]
based on https://github.com/symfony/polyfill/pull/197#discussion_r342492555 (and my answer just after), it is not, as the error handling is not the same in the polyfill and in the original extension.

**Comment by lyrixx** [comment]
OK, I removed all the type hints and codes incompatible with PHP > 5.3
Tests run on PHP 5.4+. They don't run on PHP 5.3 because the PECL does not compile.

Should be really ready now :smile:

File: src/Uuid/Uuid.php [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+    public static function uuid_create($uuid_type = UUID_TYPE_DEFAULT)
+    {
+        if (!\is_int($uuid_type)) {
+            trigger_error(sprintf("uuid_create() expects parameter 1 to be int, %s given", gettype($uuid_type)), E_USER_WARNING);
+
+            return null;
+        }
+
+        switch ($uuid_type) {
+            case UUID_TYPE_NAME:
```
Comment (jderusse) [inline src/Uuid/Uuid.php]: should be `self::UUID_TYPE_NAME` or move constantes outside the class

File: src/Uuid/Uuid.php [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+                    goto static_node;
+                }
+            }
+        } else {
+            static_node:
+            static $node;
+            if (null === $node) {
+                $node = random_int(0, 0xffffffffffff);
+            }
+        }
```
Comment (jderusse) [inline src/Uuid/Uuid.php]: Wouldn't it be faster to do avoid calling `function_exists` and `apcu_fetch` each time we generate a uuid?
```php
        static $node;
        if (null === $node) {
            if (\function_exists('apcu_fetch')) {
                $node = apcu_fetch('__symfony_uuid_node');
                if (false === $node) {
                    $node = random_int(0, 0xffffffffffff);
                    apcu_store('__symfony_uuid_node', $node);
                }
            } else {
                $node = random_int(0, 0xffffffffffff);
        }

```

File: src/Uuid/Uuid.php (hunk 10 above)
Comment (lyrixx) [reply src/Uuid/Uuid.php]: It works because the polyfill was already loaded.
But indeed, if one use the class directly (this is forbidden by `@internal` tag) it will crash.

But you are right, I will update this

File: src/Uuid/Uuid.php (hunk 11 above)
Comment (lyrixx) [reply src/Uuid/Uuid.php]: Much better (and bye bye my goto :cry: )

Comment (lyrixx) [reply src/Uuid/Uuid.php]: Note: I also fixed the support for 32-bits Sytems

File: src/Uuid/Uuid.php [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+        // not able to get a super precise timeOfDay as a unique sequence
+        $clockSeq = random_int(0, 0x3fff);
+
+        static $node;
+        if (null === $node) {
+            if (\function_exists('apcu_fetch')) {
+                $node = apcu_fetch('__symfony_uuid_node');
+                if (false === $node) {
+                    $node = sprintf('%06x%06x',
+                        random_int(0, 0xffffff),
```
Comment (xelan) [inline src/Uuid/Uuid.php]: First of all, thanks for the nice polyfill. Having UUID generation directly intergrated is great :+1:

Just a small note for the node value: According to [RFC4122](https://tools.ietf.org/html/rfc4122#section-4.1.6), the MAC address [multicast bit](https://www.geeksforgeeks.org/introduction-of-mac-address-in-computer-network/) must be set if randomly or pseudo-randomly generated node values are used.

Comment (lyrixx) [reply src/Uuid/Uuid.php]: Hello @xelan

Thanks :) 

 What does that mean exactly? What is the mylticasst bit?

Comment (xelan) [reply src/Uuid/Uuid.php]: Hi @lyrixx! Thanks for your prompt response :) According to the RFC, the UUIDs generated with random node values must be distinguishable from regular MACs via setting that bit  to 1 (LSB of first MAC octet according to the linked source). Probably it would also make sense to set the locally administered bit to 1, as the random node values are not supposed to be globally unique.

Comment (xelan) [reply src/Uuid/Uuid.php]: [Here ](https://tools.ietf.org/html/rfc4122#section-4.5)the reasoning behind that requirement is described. Based on that, I'd guess that setting both the multicast bit and the local bit would be the best solution.

Comment (lyrixx) [reply src/Uuid/Uuid.php]: done, Thanks

File: src/Util/TestListenerTrait.php [hunk 13]
```diff
@@ -109,8 +109,10 @@ function {$f['name']}{$f['signature']}
                 $mainSuite->addTest(new TestListener($suite));
             }
         }
-        foreach ($warnings as $w) {
-            $mainSuite->addTest($w);
+        if (isset($warnings)) {
```
Comment (stof) [inline src/Util/TestListenerTrait.php]: is this change still necessary, given that other changes in this file have been reverted ?

Comment (lyrixx) [reply src/Util/TestListenerTrait.php]: reverted

**Comment by nicolas-grekas** [comment]
Thank you @lyrixx.
